multi/mcxn94x-multi/cpu.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Phoenix-RTOS
+ *
+ * MCX N94x M33 CPU core driver
+ *
+ * Copyright 2025 Phoenix Systems
+ * Author: Daniel Sawka
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/file.h>
+#include <sys/interrupt.h>
+#include <sys/minmax.h>
+#include <sys/msg.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+
+#include <libtty-lf-fifo.h>
+
+#include "common.h"
+#include "dev.h"
+
+#define REG_IRQ_OFFSET 4
+/* clang-format off */
+enum {
+	reg_irq = 0, reg_irq_set, reg_irq_clr, mutex = 62
+};
+/* clang-format on */
+
+#define REG_IRQ(CPUID, REG) (*(common.mailbox + (REG_IRQ_OFFSET * (CPUID)) + (REG)))
+
+
+static struct {
+	uint8_t rxBuf[64];
+	lf_fifo_t rxFifo;
+	handle_t rxLock;
+	handle_t txLock;
+	handle_t irqLock;
+	handle_t rxCond;
+
+	volatile uint32_t *mailbox;
+	unsigned int major;
+	uint8_t coreidSelf;
+	uint8_t coreidRemote;
+} common;
+
+
+static int mailbox_irqHandler(unsigned int n, void *arg)
+{
+	size_t consumed = 0;
+	uint32_t value = REG_IRQ(common.coreidSelf, reg_irq);
+
+	if (value != 0) {
+		REG_IRQ(common.coreidSelf, reg_irq_clr) = value;
+
+		size_t rxLen = (value >> 24) & 0x3;
+
+		for (; consumed < rxLen; consumed++) {
+			uint8_t byte = (value >> (8 * consumed)) & 0xff;
+
+			/* TODO: Data is overwritten if rx fifo is full. Consider disabling
+			the interrupt somehow and leaving the value in the register until
+			the application reads from rx fifo. Also, fifo needs to be more
+			generic and at least have function to get its current length. */
+			lf_fifo_push(&common.rxFifo, byte);
+		}
+	}
+
+	return (consumed > 0) ? 1 : -1;
+}
+
+
+static ssize_t fifoRead(unsigned char *buf, size_t size, bool blocking)
+{
+	size_t bytesRead = 0;
+
+	if (buf == NULL || size == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple readers */
+	mutexLock(common.rxLock);
+
+	while (bytesRead < size) {
+		mutexLock(common.irqLock);
+		while (lf_fifo_empty(&common.rxFifo) != 0) {
+			if (!blocking) {
+				mutexUnlock(common.irqLock);
```

Comment:
Sorry for a buggy suggestion 
```suggestion
				mutexUnlock(common.irqLock);
				mutexUnlock(common.rxLock);
```

Reply:
Fixed